src/packages/plan/plan.ts

```diff
@@ -339,7 +332,293 @@ export async function importPlan(req: Request, res: Response) {
+    res.status(500);
+    res.send((error as Error).message);
+  }
+}
+
+function profileHasSegments(profileSets: ProfileSets): boolean {
+  const profileKeys = Object.keys(profileSets);
+  for (let i = 0; i < profileKeys.length; i++) {
+    if (profileSets[profileKeys[i]].segments.length) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+function getSegmentByteSize(segment: ProfileSegment): number {
+  return Buffer.byteLength(JSON.stringify(segment));
+}
+
+async function uploadDataset(req: Request, res: Response) {
+  const authorizationHeader = req.get('authorization');
+
+  const {
+    headers: { 'x-hasura-role': roleHeader, 'x-hasura-user-id': userHeader },
+  } = req;
+
+  const { body, file } = req;
+  const { plan_id: planIdString, simulation_dataset_id: simulationDatasetIdString } = body as UploadPlanDatasetPayload;
+
+  const headers: HeadersInit = {
+    Authorization: authorizationHeader ?? '',
+    'Content-Type': 'application/json',
+    'x-hasura-role': roleHeader ? `${roleHeader}` : '',
+    'x-hasura-user-id': userHeader ? `${userHeader}` : '',
+  };
+
+  let createdDatasetId: number | undefined;
+
+  try {
+    const planId: number = parseInt(planIdString);
+    const simulationDatasetId: number | undefined =
+      simulationDatasetIdString != null ? parseInt(simulationDatasetIdString) : undefined;
+    const matches = file?.originalname?.match(/\.(?<extension>\w+)$/);
+
+    if (file && matches != null) {
+      const { groups: { extension = '' } = {} } = matches;
+
+      logger.info(`POST /uploadDataset: Uploading plan dataset`);
+
+      let uploadedPlanDataset: UploadPlanDatasetJSON;
+      switch (extension) {
+        case 'json':
+          uploadedPlanDataset = await parseJSONFile<UploadPlanDatasetJSON>(file);
+          break;
+        case 'csv':
+        case 'txt': {
+          const parsedCSV: string[][] = [];
+          await new Promise((resolve, reject) => {
+            const parser = parse({
+              delimiter: ',',
+            });
+
+            parser.on('readable', () => {
+              let record;
+              while ((record = parser.read()) !== null) {
+                parsedCSV.push(record);
+              }
+            });
+            parser.on('error', error => {
+              reject(error);
+            });
+            parser.on('end', () => {
+              resolve(parsedCSV);
+            });
+
+            const fileStream = Readable.from(file.buffer);
+            fileStream.pipe(parser);
+          });
+
+          // Keep track of the time column's index separately since the name of the column is static
+          let timeColumnIndex = -1;
+
+          // Create a lookup for the profile name's index in each CSV row
+          const headerIndexMap: Record<string, number> = parsedCSV[0].reduce(
+            (prevHeaderIndexMap: Record<string, number>, header: string, headerIndex: number) => {
+              if (new RegExp(timeColumnKey).test(header)) {
```

Comment:
One thing to note here is that this will loosely match any header that contains the **substring** `time_utc`, eg. `event_time_utc` will match. I think this is **OK** here since there shouldn't be any ambiguity, since there should only be one timestamp column & the others should be numeric, but worth calling out & considering if it should be a strict equality match